source/mainnet/smart-contracts/tutorials/piggy-bank/testing.rst

```diff
@@ -281,131 +503,224 @@ The second test becomes:
 
-   let owner = AccountAddress([0u8; 32]);
-   ctx.set_owner(owner);
+Invoke the ``view`` function underneath the update:
 
-.. note::
+.. code-block:: rust
 
-   You created the account address using an array of 32 bytes, which is
-   how account addresses are represented on the Concordium blockchain.
-   These byte arrays can also be represented as a base58check encoding, but for
-   testing it is usually more convenient to specify addresses directly in bytes.
+   // .. Imports, constants, and other functions omitted for brevity.
+   #[test]
+   fn test_smash_intact(){
+
+       // .. Lines omitted for brevity.
+
+       let invoke = chain
+           .contract_invoke(
+               ACC_ADDR_OWNER,
+               Address::Account(ACC_ADDR_OWNER),
+               Energy::from(10000),
+               UpdateContractPayload {
+                   amount: Amount::zero(),
+                   address: initialization.contract_address,
+                   receive_name: OwnedReceiveName::new_unchecked("PiggyBank.view".to_string()),
+                   message: OwnedParameter::empty(),
+               },
+           )
+           .expect("Invoking `view` should always succeed");
+   }
 
-Next, set the sender to be the same address as the owner using |set_sender|_.
-Since the sender can be a contract instance as well, you must wrap the owner
-address in the |Address|_ type:
+Next step is to use the return value from (``invoke.return_value``), which is a byte array representing a tuple of ``PiggyBankState`` and ``Amount``.
+While it is possible to make assertions about the bytes directly, it is preferable to deserialize the bytes into structured Rust types.
+There is a helper method |from_bytes|_ for this exact purpose:
 
 .. code-block:: rust
 
-   let sender = Address::Account(owner);
-   ctx.set_sender(sender);
+   // .. Imports, constants, and other functions omitted for brevity.
+   #[test]
+   fn test_smash_intact(){
+
+       // .. Lines omitted for brevity.
+       let (state, balance): (PiggyBankState, Amount) =
+           from_bytes(&invoke.return_value).expect("View should always return a valid result");
 
-Lastly, you need to create a |TestHost|_ with the state and set the balance of the piggy bank
-instance using |set_self_balance|_.
+Since deserialization might fail, |from_bytes|_ returns a ``Result`` that is unwrapped here.
+If you run ``cargo test`` at this point, the compiler will complain about ``PiggyBankState`` being undeclared.
+This is because types and functions in Rust are private by default.
+To make the ``PiggyBankState`` public, edit the ``lib.rs`` file and add the ``pub`` keyword.
 
 .. code-block:: rust
+   :emphasize-lines: 1
+
+   pub enum PiggyBankState {
+      Intact,
+      Smashed,
+   }
 
-   let mut host = TestHost::new(PiggyBankState::Intact, TestStateBuilder::new());
-   let balance = Amount::from_micro_ccd(100);
-   host.set_self_balance(balance);
+.. note::
```

Comment:
Maybe this could be more prominent. Do we have something like `warn` or `important` instead?

Reply:
Good idea. I've made it important and I also added it as the fifth step in the overview in the guide.